kotlin/android/app/src/main/java/dev/firezone/android/core/data/PreferenceRepositoryImpl.kt

```diff
@@ -48,24 +48,26 @@ internal class PreferenceRepositoryImpl
 
         override fun saveToken(value: String): Flow<Unit> =
             flow {
+                val nonce = sharedPreferences.getString(NONCE_KEY, "").orEmpty()
                 emit(
                     sharedPreferences
                         .edit()
-                        .putString(TOKEN_KEY, value)
+                        .putString(TOKEN_KEY, nonce.plus(value))
                         .apply(),
                 )
             }.flowOn(coroutineDispatcher)
 
-        override fun validateCsrfToken(value: String): Flow<Boolean> =
+        override fun validateState(value: String): Flow<Boolean> =
             flow {
-                val token = sharedPreferences.getString(CSRF_KEY, "").orEmpty()
-                emit(MessageDigest.isEqual(token.toByteArray(), value.toByteArray()))
+                val state = sharedPreferences.getString(STATE_KEY, "").orEmpty()
+                emit(MessageDigest.isEqual(state.toByteArray(), value.toByteArray()))
```

Comment:
Is the MessageDigest here used as a constant-time comparison? Just making sure I understand it.

Reply:
Yup!

Reply:
(for my own reference) It is indeed constant-time. For some reason I thought it was doing a hash internally, it's probably not since hashes would be more like random time (and would probably need a salt) https://developer.android.com/reference/java/security/MessageDigest#isEqual(byte[],%20byte[])